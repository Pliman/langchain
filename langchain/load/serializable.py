from abc import ABC
from typing import Any, Dict, List, Literal, TypedDict

from pydantic import BaseModel, Field


class BaseSerialized(TypedDict):
    lc: int
    id: List[str]


class SerializedConstructor(BaseSerialized):
    type: Literal["constructor"]
    kwargs: Dict[str, Any]


class SerializedSecret(BaseSerialized):
    type: Literal["secret"]


class SerializedNotImplemented(BaseSerialized):
    type: Literal["not_implemented"]


class Serializable(BaseModel, ABC):
    @property
    def lc_namespace(self) -> List[str]:
        """
        Return the namespace of the langchain object.
        eg. ["langchain", "llms", "openai"]
        """
        return self.__class__.__module__.split(".")

    @property
    def lc_secrets(self) -> Dict[str, str]:
        """
        Return a map of constructor argument names to secret ids.
        eg. {"openai_api_key": "OPENAI_API_KEY"}
        """
        return dict()

    @property
    def lc_attributes(self) -> List[str]:
        """
        Return a list of attribute names that should be included in the
        serialized kwargs. These attributes must be accepted by the
        constructor.
        """
        return []

    lc_kwargs: Dict[str, Any] = Field(default_factory=dict, exclude=True)

    def __init__(self, **kwargs: Any) -> None:
        super().__init__(**kwargs)
        self.lc_kwargs = kwargs

    def to_json(self) -> SerializedConstructor:
        secrets = dict()

        # Merge the lc_secrets from every class in the MRO
        for cls in self.__class__.mro():
            # Once we get to Serializable, we're done
            if cls is Serializable:
                break

            # mypy doesn't understand this, but it is correct
            secrets.update(super(cls, self).lc_secrets)  # type: ignore [arg-type]

        # Get latest values for kwargs if there is an attribute with same name
        lc_kwargs = {k: getattr(self, k, v) for k, v in self.lc_kwargs.items()}
        # Add additional attributes from lc_attributes
        lc_kwargs.update({k: getattr(self, k) for k in self.lc_attributes})

        return {
            "lc": 1,
            "type": "constructor",
            "id": [*self.lc_namespace, self.__class__.__name__],
            "kwargs": lc_kwargs if not secrets else replace_secrets(lc_kwargs, secrets),
        }

    def to_json_not_implemented(self) -> SerializedNotImplemented:
        return {
            "lc": 1,
            "type": "not_implemented",
            "id": [*self.lc_namespace, self.__class__.__name__],
        }


def replace_secrets(
    root: Dict[Any, Any], secrets_map: Dict[str, str]
) -> Dict[Any, Any]:
    result = root.copy()
    for path, secret_id in secrets_map.items():
        [*parts, last] = path.split(".")
        current = result
        for part in parts:
            if part not in current:
                break
            current[part] = current[part].copy()
            current = current[part]
        if last in current:
            current[last] = {
                "lc": 1,
                "type": "secret",
                "id": [secret_id],
            }
    return result
