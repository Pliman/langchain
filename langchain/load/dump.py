import json
from typing import Any

from langchain.load.serializable import Serializable


class LangChainJSONEncoder(json.JSONEncoder):
    def __call__(self, obj: Any) -> Any:
        if isinstance(obj, Serializable):
            return obj.to_json()

        return super().default(obj)


def dumps(obj: Any, *, pretty: bool = False) -> str:
    default = LangChainJSONEncoder()
    if pretty:
        return json.dumps(obj, default=default, indent=2)
    else:
        return json.dumps(obj, default=default)
