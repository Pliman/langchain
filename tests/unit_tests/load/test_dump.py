"""Test for Serializable base class"""

from typing import Any, Dict

import openai

from langchain.chains.llm import LLMChain
from langchain.llms.openai import OpenAI
from langchain.load.dump import dumps
from langchain.load.serializable import Serializable
from langchain.prompts.prompt import PromptTemplate


class Person(Serializable):
    secret: str

    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {"secret": "SECRET"}


def test_person(snapshot: Any) -> None:
    p = Person(secret="hello")
    assert dumps(p, pretty=True) == snapshot


def test_serialize_openai_llm(snapshot: Any) -> None:
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    assert dumps(llm, pretty=True) == snapshot


def test_serialize_llmchain(snapshot: Any) -> None:
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot


def test_serialize_llmchain_with_non_serializable_arg(snapshot: Any) -> None:
    llm = OpenAI(
        model="davinci",
        temperature=0.5,
        openai_api_key="hello",
        client=openai.Completion,
    )
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot
