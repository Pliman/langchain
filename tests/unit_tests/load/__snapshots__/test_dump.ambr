# serializer version: 1
# name: test_person
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "test_dump",
      "Person"
    ],
    "kwargs": {
      "secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "SECRET"
        ]
      }
    }
  }
  '''
# ---
# name: test_serialize_llmchain
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "llms",
          "openai",
          "OpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "prompt",
          "PromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "template": "hello {name}!"
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_llmchain_with_non_serializable_arg
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "llms",
          "openai",
          "OpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          },
          "client": {
            "lc": 1,
            "type": "not_implemented",
            "id": [
              "openai",
              "api_resources",
              "completion",
              "Completion"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "prompt",
          "PromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "template": "hello {name}!"
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_openai_llm
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "llms",
      "openai",
      "OpenAI"
    ],
    "kwargs": {
      "model": "davinci",
      "temperature": 0.5,
      "openai_api_key": {
        "lc": 1,
        "type": "secret",
        "id": [
          "OPENAI_API_KEY"
        ]
      }
    }
  }
  '''
# ---
